### PR TITLE
add EditSelection mirror functionality for strokes and images

### PR DESCRIPTION
**Pull request by redweasel**
fixes https://github.com/xournalpp/xournalpp/issues/2718
fixes https://github.com/xournalpp/xournalpp/issues/1440

Simplifies the edit selection logic. Then adds in the mirror functionality, but
just like `aspectRatio` is fixed for some objects, also `mirror` is fixed for some objects.

The handling of the corners is now linear for all cases. Before this PR it was non linear for the corners, resulting in weird behavior when the mouse is not exactly on the grabbed corner.

Also adjusted cursors to match the mirrored scale directions in the corners.

### CODE REVIEW

**Comment by tattsan** [comment]
Thanks, this is the feature I've been waiting for! I tried the PR, and find the following phenomenon.

https://user-images.githubusercontent.com/6355486/106225136-dfa04000-6227-11eb-9528-3fa41b120234.mp4

If I continue extending the reflected selection, it suddenly jumps to next page.

**Comment by rolandlo** [comment]
Wow, great work so far! 

I have noticed that snapping does not work properly while moving mirrored selections. You can test it out with a rectangle. The corner closest to the position where you grabbed the selection should get snapped to grid. It still works when the selection is not mirrored.

**Comment by redweasel** [comment]
Also found https://github.com/xournalpp/xournalpp/issues/1440 which is the same feature request. In the newest version of this patch, I also fixed an issue with the selection not scrolling the page correctly when rotated (along with the here reported issue, as they were related).

Good that I initially though this would be trivial. Otherwise I would probably not have done it.

Still working on getting clang-format work... I have clang-format-11 installed.

**Comment by tattsan** [comment]
There is also #2653 in 'Discussions'.  This discusses not continuous deformation, but -100% transformation by one operation.

**Comment by redweasel** [comment]
> There is also #2653 in 'Discussions'. This discusses not continuous deformation, but -100% transformation by one operation.

I don't think adding the buttons is in the scope of this pr. I will soon (after this merged) make another one for the mirror buttons.

**Comment by tattsan** [comment]
I've used this feature for the last few days with no issue.

> I will soon (after this merged) make another one for the mirror buttons.

Thank you so much!

**Comment by Technius** [comment]
Is this backwards compatible? If not, we'll have to defer it to after the 1.1.0 release (ETA: end of this month).

**Comment by redweasel** [comment]
Yes, I tested it and the mirrored strokes (as expected) are just strokes like normal. Even the mirrored images behave normal so they are being rendered correctly, you are just not able to "unmirror" them. I think I would like to see it in 1.1.0, but I am mostly using self compiled versions anyway.

**Comment by tattsan** [comment]
I've tested the .xopp file with Xournal++ of older version.  There's a one different behavior. 
- Select the vertically mirrored image object.
- Click the bottom handle, and pull it down.
- Then the top of the selection will be extended upwards.

If you do the same operation with the new Xournal++, the bottom of the selection will be extended downwards. Only image files show this difference. Pen strokes are normal.

**Comment by rolandlo** [comment]
The mirrored image sometime disappears from the screen after unselecting it. 
Moreover the snapping issue I mentioned above (wrong corner is snapped when moving mirrored image) is still unresolved.
It would be nice to have the mirror functionality in the 1.1.0 release, but there are several issues that need to be solved yet.

**Comment by redweasel** [comment]
@rolandlo Oh now I see. Seems like I never understood snapping enough! I fixed that snapping did entirely not work, but missed that it's different for each corner. Will fix it in a moment!

**Comment by redweasel** [comment]
I fixed it. It is now using the correct corner when translating the mirrored selection. I did however find some more unstable snapping issues not related to this PR while testing. Will report them now.

**Comment by redweasel** [comment]
The fix for image rendering was a bit harder. I introduced getters for `getVisualX/Y/Width/Height` into `Element`. That looked like the most reliable way to do this. Hope I didn't miss a spot to swap the normal getter for the visual getter.

I have to retract my claim that this is backwards compatible. It is almost backward compatible, but images mirrored images will only render on full page rerenders in older versions.

**Comment by redweasel** [comment]
This should be ready for a final review. I disabled mirror for everything except for strokes for now.

File: src/control/tools/EditSelection.cpp
Comment (rolandlo) [inline src/control/tools/EditSelection.cpp:459]: I thinks this needs correction. You will see the difference if your selection consists of an ellipse with a large eccentricity.
```suggestion
            f = std::copysign(std::max(std::abs(f), minSize / std::min(std::abs(this->width), std::abs(this->height))), f);
```

File: src/control/tools/EditSelection.cpp
Comment (rolandlo) [inline src/control/tools/EditSelection.cpp:691]: Could you please define `constexpr` for these magical numbers 20 and 8 which are used (here and elsewhere) for the position of the delete and rotate rectangles? That would make it more readable.

**Review by rolandlo — CHANGES_REQUESTED** [review]
Except of the two minor change requests I made everything seems to work very well. I have tested copying/pasting, cloning, undo/redo, save/load, snapping, rotation, resizing and moving. For undo/redo to work properly I just needed to rebase on current master.
I love your work, both for the mirror capability (which I will use in geometry class often) and for the simplifications in the code.

So please make the two requested changes and rebase on current master.

**Review by rolandlo — APPROVED** [review]
Ok, great! Thanks for this contribution. Do you want to squash any commits (or make other changes) before I merge it?

**Comment by redweasel** [comment]
Just to be sure, what is the general rule around here for commits in a PR? In other projects it was always one commit per feature. I would squash the commits If that's preferred here.

**Comment by rolandlo** [comment]
We often have only one commit per feature. If the commits (of a larger PR) are nicely separated and only take care of one problem at a time we will accept multiple commits. Fixups of previous commits like "Apply clang format" or "Fix the thing I forgot previously" should definitely get squashed away. I think for this PR it would make sense to squash it all into one commit. If you would prefer to have the fix of automatic scroll separated from the rest, this would probably justify to squash the PR into two commits. 
Note that we have sometimes forgotten to squash the commits before we merged it. So if you look at the commit history you will find exceptions to what I just wrote.

**Comment by redweasel** [comment]
wait a sec, do not merge now. I made a mistake!

**Comment by redweasel** [comment]
Ok the PR is ready for merge now.

**Comment by rolandlo** [comment]
Thanks again. Great job!

**Comment by danielgomez3** [comment]
I'm not sure how to flip my selection in xournalpp.. apparently this feature was added? I'm using Xournalpp 1.2.3 on NixOS.

**Comment by rolandlo** [comment]
> I'm not sure how to flip my selection in xournalpp.. apparently this feature was added? I'm using Xournalpp 1.2.3 on NixOS.

For what was implemented in this PR, see the screen recording https://github.com/xournalpp/xournalpp/pull/2723#issuecomment-769538152

There is a Lua plugin for flipping selections of strokes and a Lua plugin for flipping images. What are you interested in?

**Pull request merged**
